### PR TITLE
ping device simulation

### PR DESCRIPTION
**Pull request by jaxxzer**

### CODE REVIEW

File: brping/ping1d-simulation.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        self._profile_data_length = 200 # number of data points in profile messages (constant for now)
+        self._pulse_duration = 100 # length of acoustic pulse (constant for now)
+        self._ping_number = 0 # count the total measurements taken since boot
+        self._ping_interval = 100 # milliseconds between measurements
+        self._mode_auto = True # automatic gain and range selection
+        self._mode_continuous = True # automatic continuous output of profile messages
+
+    # read incoming client data
+    def read(self):
+            try:
```
Comment (Williangalvani) [inline brping/ping1d-simulation.py]: there is an extra indentation level here

File: brping/ping1d-simulation.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    def write(self, data):
+        if self.client is not None:
+            print("sending data to client")
+            self.sockit.sendto(data, self.client)
+
+    # Send a message to the client, the message fields are populated by the
+    # attributes of this object (either variable or method) with names matching
+    # the message field names
+    def sendMessage(self, message_id):
+            msg = PingMessage(message_id)
```
Comment (Williangalvani) [inline brping/ping1d-simulation.py]: another extra indentation level

File: brping/ping1d-simulation.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+            # pull attributes of this class into the message fields (they are named the same)
+            for attr in payload_dict[message_id]["field_names"]:
+                try:
+                    # see if we have a function for this attribute (dynamic data)
+                    # if so, call it and put the result in the message field
+                    setattr(msg, attr, getattr(self, attr)())
+                except Exception as e:
```
Comment (Williangalvani) [inline brping/ping1d-simulation.py]: maybe specify this Exception?

File: brping/ping1d-simulation.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        if self._mode_auto:
+            self._scan_length = self.periodicFnInt(2000, 3000, 0.2)
+        return self._scan_length
+
+    def profile_data(self):
+        stops = 20
+        len = int(200/stops)
+        data = []
+        for x in range(stops):
+            data = data + [int(x * 255/stops)] * len
```
Comment (Williangalvani) [inline brping/ping1d-simulation.py]: space around operators looks inconsistent. Did you mean 255/stops is the step?

File: brping/ping1d-simulation.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
+# This script simulates a Blue Robotics Ping Echosounder device
```
Comment (patrickelectric) [inline brping/ping1d-simulation.py]: Missing shebang

**Comment by patrickelectric** [comment]
```
 ✘ patrick@pichau  ~/git/blue/ping-python/brping  ➦ 979e500 ●  pip install bluerobotics-ping --user
Requirement already satisfied: bluerobotics-ping in /home/patrick/.local/lib/python3.7/site-packages (0.0.2)
Requirement already satisfied: pyserial in /usr/lib/python3.7/site-packages (from bluerobotics-ping) (3.4)
 patrick@pichau  ~/git/blue/ping-python/brping  ➦ 979e500 ●  python -m ping1d-simulation 
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/patrick/git/blue/ping-python/brping/ping1d-simulation.py", line 162, in <module>
    sim.sendMessage(PING1D_PROFILE)
  File "/home/patrick/git/blue/ping-python/brping/ping1d-simulation.py", line 76, in sendMessage
    msg.pack_msg_data()
  File "/home/patrick/.local/lib/python3.7/site-packages/brping/pingmessage.py", line 471, in pack_msg_data
    self.msg_data = bytearray(struct.pack(msg_format, *values))
struct.error: pack expected 6 items for packing (got 15)
 ✘ patrick@pichau  ~/git/blue/ping-python/brping  ➦ 979e500 ●  

```

**Comment by jaxxzer** [comment]
@patrickelectric use python3

**Comment by jaxxzer** [comment]
> Also, missing executable permission.

```
$ ls -l brping/
total 108
-rw-r--r-- 1 jack jack    89 Jan 19 12:31 __init__.py
-rw-r--r-- 1 jack jack   211 Jan 24 15:51 __init__.pyc
-rw-r--r-- 1 jack jack 32470 Jan 25 12:22 ping1d.py
-rw-r--r-- 1 jack jack 13607 Jan 24 15:51 ping1d.pyc
-rw-r--r-- 1 jack jack  5975 Jan 28 14:29 ping1d-simulation.py
-rw-r--r-- 1 jack jack 22945 Jan 25 17:19 pingmessage.py
-rw-r--r-- 1 jack jack 13322 Jan 24 15:51 pingmessage.pyc
-rw-r--r-- 1 jack jack  3413 Jan 19 12:31 pingproxy.py
```

The permission is the same as everything else.

**Comment by patrickelectric** [comment]
@jaxxzer 	I'm using.

**Comment by jaxxzer** [comment]
@patrickelectric This requires https://github.com/bluerobotics/ping-protocol/pull/118 merge

**Comment by patrickelectric** [comment]
Ok, this should be tagged in the PR.

**Comment by jaxxzer** [comment]
I forgot. Now it is.

**Comment by jaxxzer** [comment]
> But ping1d-simulation.py is a program by itself, this one should have +x now ?

https://github.com/bluerobotics/ping-python/issues/43

File: brping/ping1d-simulation.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/python -u
```
Comment (patrickelectric) [inline brping/ping1d-simulation.py]: You should use env and noy python directly.

Comment (patrickelectric) [reply brping/ping1d-simulation.py]: Python3 only
```
#!/usr/bin/env python3
```
or
Python2 only
```
#!/usr/bin/env python2
```
or
Both python versions
```
#!/usr/bin/env python
```

**Comment by jaxxzer** [comment]
> shebang should use /usr/bin/env and not python directly

This matches the case in the rest of the repository.
 #44

**Comment by jaxxzer** [comment]
> shebang should use /usr/bin/env and not python directly

done

**Pull request merged**